install.sh

```diff
@@ -1,41 +1,57 @@
 #!/usr/bin/env bash
+
 set -e
 
-resolve_link() {
-  $(type -p greadlink readlink | head -n1) "$1"
+PREFIX="$1"
+if [ -z "$1" ]; then
+  { echo "usage: $0 <prefix>"
+    echo "  e.g. $0 /usr/local"
+  } >&2
+  exit 1
+fi
+
+export BATS_READLINK='true'
+if command -v 'greadlink' >/dev/null; then
+  BATS_READLINK='greadlink'
+elif command -v 'readlink' >/dev/null; then
+  BATS_READLINK='readlink'
+fi
+
+bats_resolve_link() {
+  if ! "$BATS_READLINK" "$1"; then
+    return 0
+  fi
 }
 
-abs_dirname() {
-  local cwd="$(pwd)"
+bats_absolute_path() {
```

Comment:
As above, also wonder if this should be refactored into a shared file as it's reused

Reply:
I was thinking this through, and still am, but there's a chicken-and-egg problem—or there might not be, and I'd be interested in your thoughts.

Apologies in advance for the length.

The way the system works depends upon `bin/bats` discovering the real Bats root directory. For example, if Bats is installed at `/opt/bats`, then `/opt/bats/bin/bats` needs to find `/opt/bats/libexec/bats`. `/opt/bats/libexec/bats` then adds `/opt/bats/libexec` to the front of `PATH`, so then Bats can find the rest of the `/opts/bats/libexec` scripts. Because of this, `/usr/local/bin/bats` can symlink to `/opt/bats/bin/bats`, and the rest of Bats's internals can remain encapsulated within `/opt/bats` such that they don't pollute the `/usr/local` namespace (the man pages notwithstanding, unless `/opt/bats/man` is added to `MANPATH` somehow).

Previously, having `bin/bats` be a symlink to `../libexec/bats` worked fine in general—but broke down in the `/usr/local/bin/bats` with `/opt/bats` scenario I described above, and was creating complications for some Windows-based platforms. Hence why I made `bin/bats` a normal script, then pushed the real path logic in there.

Now `install.sh` from sstephenson/bats also originally contained this real path-finding logic. Having duplicate code rubs me terribly wrongly, but if `bin/bats` needs to find its real path, and `install.sh` needs to find its real path...then how do they each find their real path so that they can import a shared script containing the real path logic?

One possibility I thought of, but I'm not sure it's worth the complication: If there were more than one Bats executable, they could all source `${0%/*}/bats`, which could use `${BATS_SOURCE[1]}` to discover which script sourced it, and could then dispatch logic between them. We could set up `install.sh` to source `${0%/*}/bin/bats` to achieve this—but then we presume `install.sh` isn't symlinked directly, which would land us back in the same predicament.

But what if we assumed (or specified that) `install.sh` itself won't (shouldn't) ever be symlinked? In that case we probably _could_ just compute the Bats root directory as the parent of `install.sh` (`${0%/*}`) without finding the real path. In other words, we can't assume `/usr/local` is the correct Bats root for `/usr/local/bin/bats`, but we may be safe in assuming (or specifying) that `/opt/bats` is the correct Bats root for `/opt/bats/install.sh`. This would eliminate the need for the duplicated code in the latter script.

Thoughts?

Reply:
Hehe, I've been mulling it over and hacking around. Will pull the branch and check it out, cheers.

Reply:
Sorry to cause more problems (!) - in writing the usage guide in #94 I found an edge case, which I've put into the travis build as per #93. Essentially the changes in #92 break running the tests when the tests are mounted into the container at a path that isn't a direct decedent of the bats base directory.

Current master doesn't fail under these conditions. I don't suspect anybody except us is using the pattern described above, but I can't be sure of the repercussions as I'm not as close to this code as you are,

Reply:
Since it's the unofficial Bats strict mode test that's failing, I think I need to add a couple `declare` statements back into `install.sh` and that should do the trick. Will try in a couple hours.

Reply:
Wait, that might not be it...either way, sure it's something small, and I'm doing a terrible job multitasking right now. :-P

Reply:
To close the loop on this: I found and explained the breakage in #93, which actually wasn't due to #92. The "unofficial Bash strict mode" test failed because it was looking for `bats` relative to `BATS_TEST_DIRNAME`, which it couldn't because the `test` directory was mounted into the container in isolation. Mounting the root of the Bats project resolved the issue.